### DISCUSSION TITLE
Sign-up database trigger to insert into public users table

### DISCUSSION THREAD

**Discussion by skoshy** [supabase/supabase#306] [Questions] [answered]
Hi, I'm trying to set up a trigger to automatically copy new users from `auth` to my public `users` table. I've drafted up the below, but I get a database error every time I try to signup a new user (`"Database error saving new user"`). Any ideas on what could be the issue?

```sql
CREATE OR REPLACE FUNCTION auth.signup_copy_to_users_table()
RETURNS TRIGGER LANGUAGE plpgsql AS $function$
  BEGIN
    INSERT INTO public.users(id)
    VALUES(NEW.id);
  
    RETURN NEW;
  END;
$function$;

DROP TRIGGER IF EXISTS signup_copy on auth.users;
CREATE TRIGGER signup_copy
AFTER INSERT OR UPDATE ON auth.users
FOR EACH ROW EXECUTE PROCEDURE signup_copy_to_users_table();
```

In general it'd be useful to be able to keep the auth/public users tables in sync with the `id` and `email` fields at least.

**Comment by stfnsr** [comment]
Same issue here.
Interestingly, there's no problem if i e.g. update the auth.users table via the trigger function, but everything in the public schema is off-limits somehow.

**Comment by kiwicopple** [comment, marked as answer]
I suspect it's permissions - the trigger is probably getting called by the `anon` user. You can try to add `security definer` to the function, which will use the same permissions as the user who created the function.

Here is a snippet that should work:

```sql

/**
* This trigger automatically creates a user entry when a new user signs up via Supabase Auth.
*/ 
create function public.handle_new_user() 
returns trigger as $$
begin
  insert into public.users (id)
  values (new.id);
  return new;
end;
$$ language plpgsql security definer;
create trigger on_auth_user_created
  after insert on auth.users
  for each row execute procedure public.handle_new_user();
```

**Reply by skoshy** [reply]
That worked! Thank you so much! Here's my final script in case anyone needs it.

```sql
CREATE OR REPLACE FUNCTION signup_copy_to_users_table()
RETURNS TRIGGER AS $$
  BEGIN
    INSERT INTO public.users (id, email)
    VALUES(new.id, new.email);
  
    RETURN NEW;
  END;
$$
LANGUAGE plpgsql SECURITY DEFINER;

DROP TRIGGER IF EXISTS signup_copy on auth.users;
CREATE TRIGGER signup_copy
AFTER INSERT ON auth.users
FOR EACH ROW EXECUTE PROCEDURE signup_copy_to_users_table();
```

**Reply by alfredosalzillo** [reply]
I have tried this solution but I'm still having an error 

```json
{"code":500,"msg":"Database error saving new user","error_id":"9b354033-3bb8-4c5d-af3c-82f67695373c"}
```

**Reply by silentworks** [reply]
@alfredosalzillo since this was already answered can you open a new discussion around the issue you are having with the steps you took i.e. code example.

**Reply by lucasnantonio** [reply]
> I have tried this solution but I'm still having an error
> 
> ```json
> {"code":500,"msg":"Database error saving new user","error_id":"9b354033-3bb8-4c5d-af3c-82f67695373c"}
> ```

I am also getting this, is there a link to this issue and its resolution?

**Reply by romlytech** [reply]
Same issue here.

Anyone else have some ideas?

**Reply by rotimi-best** [reply]
I also have this same issue.

**Reply by chrisLoPresti** [reply]
same issue

**Reply by rotimi-best** [reply]
I ended up manually creating a profile record if it doesn't exist when the user enters the application.

**Reply by inian** [reply]
For anyone coming along, remember to turn on RLS and to disable realtime on your public.users table. If not, anyone with an anonkey will be able to access it.

**Reply by fellowshipdevs** [reply]
> > I have tried this solution but I'm still having an error
> > ```json
> > {"code":500,"msg":"Database error saving new user","error_id":"9b354033-3bb8-4c5d-af3c-82f67695373c"}
> > ```
> 
> I am also getting this, is there a link to this issue and its resolution?

Anyone still getting this, make sure you don't have non-nullable fields in the table. Spent an hour trying to figure out why my use case wasn't working, and it was because of that.

**Reply by NixBiks** [reply]
That worked for me. But it doesn't work if I have the following update trigger as well.

```sql
-- update a row in public."Users" when the email is updated
create or replace function public.handle_updated_user()
returns trigger as $$
begin
  update public."Users"
  set email = new.email
  where id = new.id;
  return new;
end;
$$ language plpgsql security definer;

-- trigger the function every time a user is updated
drop trigger if exists on_auth_user_updated on auth.users;
create trigger on_auth_user_updated
  after update on auth.users
  for each row execute procedure public.handle_updated_user();
```


 I can't delete users either if I have this delete trigger

```sql
-- delete a row from public."Users" when the user is deleted
create or replace function public.handle_deleted_user()
returns trigger as $$
begin
  delete from public."Users" where id = old.id;
  return old;
end;
$$ language plpgsql security definer;

-- trigger the function every time a user is deleted
drop trigger if exists on_auth_user_deleted on auth.users;
create trigger on_auth_user_deleted
  after delete on auth.users
  for each row execute procedure public.handle_deleted_user();
```

Anyone able to spot the issues?

**Reply by NixBiks** [reply]
My issue was that public."Users".id was `TEXT` and not `UUID`. So I had to cast. Note the `::text` in the following working example 

```sql
-- update a row in public."Users" when the email is updated
create or replace function public.handle_updated_user()
returns trigger as $$
begin
  update public."Users"
  set email = new.email
  where id = new.id::text;
  return new;
end;
$$ language plpgsql security definer;

-- trigger the function every time a user is updated
drop trigger if exists on_auth_user_updated on auth.users;
create trigger on_auth_user_updated
  after update on auth.users
  for each row execute procedure public.handle_updated_user();


-- delete a row from public."Users" when the user is deleted
create or replace function public.handle_deleted_user()
returns trigger as $$
begin
  delete from public."Users" where public."Users".id = old.id::text;
  return old;
end;
$$ language plpgsql security definer;

-- trigger the function every time a user is deleted
drop trigger if exists on_auth_user_deleted on auth.users;
create trigger on_auth_user_deleted
  after delete on auth.users
  for each row execute procedure public.handle_deleted_user();
```

**Reply by aditodkar** [reply]
For me it is still not working. I have created public.users table and added id with foreign key relation to auth.users and id is also primary key in public.users. I have also enabled anon role policy but still getting database sign up error. Any idea what I am missing?
code:
```
/**
* This trigger automatically creates a user entry when a new user signs up via Supabase Auth.
*/ 
create function public.handle_new_user() 
returns trigger as $$
begin
  insert into public.users (id)
  values (new.id);
  return new;
end;
$$ language plpgsql security definer;
create trigger on_auth_user_created
  after insert on auth.users
  for each row execute procedure public.handle_new_user();

```

**Reply by NixBiks** [reply]
> For me it is still not working. I have created public.users table and added id with foreign key relation to auth.users and id is also primary key in public.users. I have also enabled anon role policy but still getting database sign up error. Any idea what I am missing?
> 
> code:
> 
> ```
> 
> /**
> 
> * This trigger automatically creates a user entry when a new user signs up via Supabase Auth.
> 
> */ 
> 
> create function public.handle_new_user() 
> 
> returns trigger as $$
> 
> begin
> 
>   insert into public.users (id)
> 
>   values (new.id);
> 
>   return new;
> 
> end;
> 
> $$ language plpgsql security definer;
> 
> create trigger on_auth_user_created
> 
>   after insert on auth.users
> 
>   for each row execute procedure public.handle_new_user();
> 
> 
> 
> ``` 

How does your `public.users` table definition look like?

**Reply by aditodkar** [reply]
@mr-bjerre public.users is a table with only 1 column i.e id which is primary key and also has foreign key mapping to auth.users id. RLS is enabled and I have also added policies with role='anon'

**Reply by NixBiks** [reply]
But what's the type of id?

**Reply by aditodkar** [reply]
I deleted trigger and re-run it now what is happening sign up is working fine but trigger is not getting executed. Also after re-run it is saying `function "handle_new_user" already exists with same argument types` but in triggers section it does not show any on_auth_user_created trigger but in functions I see handle_new_user function which I created.

**Reply by aditodkar** [reply]
id type is uuid

**Reply by NixBiks** [reply]
> I deleted trigger and re-run it now what is happening sign up is working fine but trigger is not getting executed. Also after re-run it is saying `function "handle_new_user" already exists with same argument types` but in triggers section it does not show any on_auth_user_created trigger but in functions I see handle_new_user function which I created.
> 
> 

You can do `create or replace function` instead of `create function`. It won't create a trigger if creating the function throws an error.

**Reply by NixBiks** [reply]
Also I guess an issue could be that you have a relation to auth.users. You are kinda making a chicken or egg paradox - ie. you can't add a relation to a user that is not yet created. I'm not sure about this though 🤔

**Reply by aditodkar** [reply]
I checked with foreign key relation to auth.users id it works fine. Thanks

**Reply by younusmahmood** [reply]
I'm running into this same issue, has anyone figured it out? 

I tried all the examples I could find in these github threads. Created the public.users table per docs too. It was actually working at one point but stopped recently so not sure what changed. I moved away from this for now and just making another api call in my app to create the user in the public.users table after the supabase signup.

Here's my code:
```
CREATE FUNCTION
  public.handle_new_user()
  RETURNS TRIGGER AS
  $$
  BEGIN
    INSERT INTO public.users (id)
    VALUES (NEW.id);
    RETURN NEW;
  END;
  $$ LANGUAGE plpgsql SECURITY DEFINER;
CREATE TRIGGER on_auth_user_created
    AFTER INSERT
    ON auth.users
    FOR EACH ROW
    EXECUTE FUNCTION public.handle_new_user();
```

and here are the policies on my user table:

Public users are viewable by everyone.
true
SELECT
Edit

Delete
Users can insert their own profile.
(uid() = id)
INSERT
Edit

Delete
Users can update own profile.
(uid() = id)

**Reply by Donnerstagnacht** [reply]
I hat the same issue. In my case I added a"unique" column in profile table, but I did not insert a unique name during signUp. I removed the unique constraint to solve it.

A good strategy to approach this error might be to check all parameters which might stop a function from inserting a new row -like security definer, rls, unique constraints, nullable constraints...

**Reply by activenode** [reply]
We had basically the same as Paul and then we tried the exact script from @kiwicopple but neither of those seem to work. What seems to work though is defining the same thing but with the function being part of `auth.` schema. Is there something from PGv15 that could cause this?

Also if you do so it works but the function doesn't show in the functions UI as they seem to not show funcs from other schemas.

**Reply by kiwimagic** [reply]
Adding my 2c worth. After days trying to create a parallel auth mechanism between firebase phone sign in and supabase the answer above "saved" me from further frustration - adding the magic " security definer;" after my function allowed me to do what I needed to create a user fully via postgres + API. Unfortunately there are so many people struggling with Auth that this answer is buried in the masses.  As it is 3 years after this was first asked and still an issue (and not [clearly?] explained yet in the documentation) I will summarize my findings here so they may be helpful for someone in the future. 

1.  can't touch/don't touch the auth.users table - accept what supabase has/does with this - put your data in public.[own user table] name[s] and add FK to auth.users
2.  you can access "create user" signup via api [...]supabase.co/auth/v1/signup (POST) {this is very buried in discussion threads}
3.  you can send custom data to be held in the auth.users jsonb meta with variables (this is from my FlutterFlow API call definition Body JSON){
  "email": "<email>",
  "password": "<password>",
  "data": {
    "app_source": "<app_source>",
    "zone": "<zone>",
    "phone": "<phone>"
  }
}
4. modify the scripting in the Function called public.handle_new_user() to do what you want data wise to tables in your PUBLIC schema eg create your own "user_profile" table(s) and link the auth.users id to a uuid (may or may not be PK depending on how you architect your public.user_profile schema) NOTE there is sample code provided in the Supabase SQL Editor in a script called User Management Starter (this appeared for me in one of the 2023 updates review this it's a good starter AND DOES have the "security definer" if I had started with this I would have been fine but I copied out the function code from the function UI and reworked my own without seeing this )
5. ***KEY*** at the end of your modified public.handle_new_user() trigger function make sure you add the **security definer** 
_example framework:_ 
```
CREATE OR REPLACE FUNCTION public.handle_new_user() RETURNS TRIGGER AS $$
DECLARE
...
BEGIN
...
do your public.[table] insert/update ..etc here
...
RETURN NEW; 

END;
$$ language plpgsql security definer; -- THESE 2 WORDS ARE THE MISSING LINK!
```
6. Note once I found the Supabase Log > Postgres [filter Severity = Error] the issue with the permission and tracking down the security definer mention in this discussion became a lot easier! Most of my time was spent trying to figure out what API to use and the syntax to call it. If you are stuck with a 500 ERROR - start with the logs! Good luck.

**Reply by hidaviddong** [reply]
it works! thanks!

**Reply by davidpedrero** [reply]
This worked for me! Thanks! Adding the security definer to the function was the key for me

**Reply by teamclouday** [reply]
Thank you it works for me too! I had a permission denied error from function that's inserting a RLS enabled public table with role `supabase_auth_admin`. Adding `security definer` solves it perfectly. I guess this role doesn't have sufficient permission.

**Reply by walifile** [reply]
> > > I have tried this solution but I'm still having an error
> > > ```json
> > > {"code":500,"msg":"Database error saving new user","error_id":"9b354033-3bb8-4c5d-af3c-82f67695373c"}
> > > ```
> > 
> > 
> > I am also getting this, is there a link to this issue and its resolution?
> 
> Anyone still getting this, make sure you don't have non-nullable fields in the table. Spent an hour trying to figure out why my use case wasn't working, and it was because of that.

I really appreciate it! I was getting the same error and couldn't fix it. This comment actually works. Thank you!

**Reply by CamiloTello002** [reply]
Thank you!

**Reply by boiwantlearncode** [reply]
Thanks, still works in 2025

**Reply by tfrank11** [reply]
I hope both sides of your pillow are cold. Tysm.

**Comment by stfnsr** [comment]
Wow, thanks @kiwicopple - this works 👍 

Would this be a useful thing to include in the docs?
E.g., in the area where creation of a public.user table is advised?

**Reply by kiwicopple** [reply]
Done! https://supabase.io/docs/guides/auth/#create-a-publicusers-table

**Reply by imjared** [reply]
heads up the link seems to fail with the trailing slash. here's this for the next google traveler :)
 https://supabase.io/docs/guides/auth#create-a-publicusers-table

**Reply by inian** [reply]
It should work with the trailing slash as well now!

**Reply by grantwparks** [reply]
I feel really dumb about this and I apologize if I'm asking in a weird place.

I can't find where I'm supposed to add triggers like this from the Supabase project console.  I've been experimenting with PostgresT for a project and wanted to try Supabase out alongside, but I don't know where to put triggers, constraints and the like from that console.

**Reply by silentworks** [reply]
No problem @grantwparks just go to the SQL editor which is like the 3 icons down from the home icon. I have attached a short screen recording of how to get to it.

https://user-images.githubusercontent.com/79497/103572617-bf998d00-4ec4-11eb-9415-cbbb3cd44a41.mp4

**Reply by grantwparks** [reply]
Ah, OK, my source code-driven thinking was "how does this get saved to the schema for rebuilding?", but if I understand better now, the DB instance figuratively just persists with whatever code has been applied up to a given time.  Thank you.

**Reply by chriserl** [reply]
Hey, is there anyway to create this function without SQL?

**Reply by silentworks** [reply]
Not that I know of at this moment in time.

**Reply by dperrera** [reply]
FYI, the docs for this seem to have been removed. There is a reference to the technique but no information on _how_ to create the function or even that the solution is a function.

<img width="835" alt="image" src="https://user-images.githubusercontent.com/226921/126884244-48c2ec25-58fd-4eba-ab33-4e9ebe5a87ee.png">

**Reply by MildTomato** [reply]
@dperrera I think you're right.

we should perhaps add a function example in [here]( https://supabase.io/docs/guides/auth/managing-user-data) to show how to auto populate a table for user data ?

**Reply by inian** [reply]
For anyone coming along, remember to turn on RLS and to disable realtime on your public.users table. If not, anyone with an anonkey will be able to access it.

**Reply by user72356** [reply]
> For anyone coming along, remember to turn on RLS and to disable realtime on your public.users table. If not, anyone with an anonkey will be able to access it.

Hey I'm coming in late, but what do you mean by that? I don't use RLS in my account and I don't have realtime enabled. Am I safe? Can you explain a little better/more details?

**Reply by GaryAustin1** [reply]
@user72356  You need RLS enabled and policies set or anyone can access (and even change) your data with the anon key which is exposed on browsers.
https://supabase.com/docs/guides/auth/row-level-security

**Reply by user72356** [reply]
OK so just to confirm, if I enable realtime publications for a table, and I disable RLS for that table, all clients connected with the anon key will receive updates.

However, if I disable realtime publications and disable RLS for a table, the anon key can't read that table in any way?

Guess I'm in that situation: https://github.com/supabase/supabase/discussions/4547

I'm surprised the DB is open to the public by default, quite a shocker to me. I was not planning on using RLS because my API is doing the heavy lifting with regular SQL queries, and I only need the supabase client for auth and subscriptions on one small table.

**Reply by GaryAustin1** [reply]
They are two different things.  Ignoring Realtime, if you have RLS disabled, anon key can read all your tables thru the REST API.
You have to Enable RLS and grant policies for authenticated users, or whatever control you want.

**Comment by TStone45** [comment]
I am unable to work out what to do in terms of overcoming the fact that I am successfully adding email first name last name JWT and password into the auth.users table but try as I might syntax errors prevent me from finishing a function, I don't even know how to set up a function that observes changes from the auth table. I need more than help I need Jesus to take the wheel.

**Reply by activenode** [reply]
I'm not jesus but what you need is a function that also listens on the updates,not just the insert. And then determine for yourself if it's done.

I do this by checking if the entries in metadata are not empty like tldr:

```
DECLARE
has_proper_metadata boolean;

BEGIN
    has_proper_metadata = EXISTS(SELECT 1 FROM auth.users a WHERE 
      (raw_app_meta_data::jsonb ->> 'foobar1') IS NOT NULL
      and
      (raw_app_meta_data::jsonb ->> 'foobar2') IS NOT NULL
    AND a.email = NEW.email);

  IF has_proper_metadata THEN
    -- do your stuff
  ELSE
    RETURN NEW; -- continue 
  END IF;
END;

```

Cheers [activeno.de](https://activeno.de)

**Reply by TStone45** [reply]
You may not be Jesus but you have forgiven me my syn…taxes and have resurrected my journey to the promised land. 🙌🏾

**Reply by activenode** [reply]
Well then, spread the word (and my YouTube channel) if I may :D

**Comment by TStone45** [comment]
Hey Boss another question: 

Raw meta data apparently isnt recognising the following public trigger:

![public function](https://github.com/supabase/supabase/assets/138574369/50f81901-887f-40e8-81f0-2d0e49d1299b)

This is the function as its written


![handlercode](https://github.com/supabase/supabase/assets/138574369/2ce76894-2f0c-4090-a966-a916899b205f)

And heres the policies for that public.users table.

![users table policies](https://github.com/supabase/supabase/assets/138574369/c43a9a97-5aa9-43d4-83bd-198a001093da)


And finally this is the sql editor query that created the public handler

![publichandlersqleditorquery](https://github.com/supabase/supabase/assets/138574369/0896a660-bb10-471d-8146-a851c9139e3b)

Why is it still not creating a record upon the user being added to auth.users
'??

**Reply by activenode** [reply]
Didn't know one can use `new` lowercase. But I guess this isn't the issue.  If this doesn't create a user entry in `public.users` then it should usually also prevent the creation of the `auth.users` itself. Have you checked the Postgres logs in the Dashboard?

**Reply by GaryAustin1** [reply]
Checking the Postgres log is the correct thing to do like activenode suggests.

If user_id column is UUID then you will at least need to cast that json value from text with ::UUID.  ->> returns the value as text.
Edit: I see you also have id, so user_id may indeed be text type.

**Reply by GaryAustin1** [reply]
Also you should pull email from new.email.  If you signup with email the email will not be in the metadata.  That only comes from OAuth signups.

**Reply by TStone45** [reply]
Thank you Gary and David really appreciate your support, 

I will try to reference and pull in the email from the auth table (or do a new.email) and do away with that particular column in the public table, 

the logs are below, I've actual been speaking to brianglass a SB guy about the triggers also not working correctly he advised me to turn on debugging, I have, 

![Brianassistancelogstriggerpublicauth](https://github.com/supabase/supabase/assets/138574369/dadf8352-9f83-4e74-93e2-18792f3e27d9)



Logs: of most recently failure. 

{"component":"api","error":"failed to close prepared statement: ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02): ERROR: column \"firstname\" of relation \"users\" does not exist (SQLSTATE 42703)","level":"error","method":"POST","msg":"500: Database error saving new user","path":"/signup","referer":"http://localhost:3000","remote_addr":"111.220.138.33","time":"2024-05-10T00:19:42Z","timestamp":"2024-05-10T00:19:42Z"}
Timestamp
2024-05-10T00:19:42.000Z

Metadata
[
  {
    "message": null,
    "timestamp": "2024-05-10T00:19:42Z",
    "__MONOTONIC_TIMESTAMP": null,
    "CODE_FUNC": null,
    "instance_id": null,
    "status": null,
    "_CMDLINE": null,
    "method": "POST",
    "_SYSTEMD_CGROUP": null,
    "CODE_FILE": null,
    "EXECUTABLE": null,
    "_EXE": null,
    "UNIT": null,
    "level": "error",
    "_COMM": null,
    "duration": null,
    "issuer": null,
    "_LINE_BREAK": null,
    "_SOURCE_REALTIME_TIMESTAMP": null,
    "msg": "500: Database error saving new user",
    "action": null,
    "login_method": null,
    "_UID": null,
    "host": "db-bzqvkknsoikjeqnonnkv",
    "PRIORITY": null,
    "_CAP_EFFECTIVE": null,
    "_PID": null,
    "INVOCATION_ID": null,
    "_SYSTEMD_UNIT": null,
    "source_type": null,
    "SYSLOG_FACILITY": null,
    "request_id": null,
    "CODE_LINE": null,
    "path": "/signup",
    "component": "api",
    "project": null,
    "user_id": null,
    "auth_event": [],
    "args": [],
    "referer": "http://localhost:3000",
    "factor_id": null,
    "provider": null,
    "client_id": null,
    "remote_addr": "111.220.138.33",
    "_SYSTEMD_SLICE": null,
    "_SYSTEMD_INVOCATION_ID": null,
    "header": null,
    "_MACHINE_ID": null,
    "_AUDIT_LOGINUID": null,
    "_TRANSPORT": null,
    "_SELINUX_CONTEXT": null,
    "MESSAGE_ID": null,
    "__REALTIME_TIMESTAMP": null,
    "metadata": [],
    "_STREAM_ID": null,
    "metering": null,
    "time": null,
    "_GID": null,
    "_BOOT_ID": null,
    "SYSLOG_IDENTIFIER": null,
    "_AUDIT_SESSION": null,
    "error": "failed to close prepared statement: ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02): ERROR: column \"firstname\" of relation \"users\" does not exist (SQLSTATE 42703)"
  }
]

Will update later today if still unresolved.

**Reply by GaryAustin1** [reply]
@TStone45 I missed it in your code.  But Postgres does not like caps.  If you use caps for names of tables or columns you MUST put them in double quotes or they go lower case.  So your error is that all your column names are converting to lower case as you did not use "lastName".   `lastName` will become `lastname` which will not match `lastName` as a column name.  `"lastName"` will match.

![image](https://github.com/supabase/supabase/assets/54564956/a901122f-12ba-4eee-bcb2-a04454da0147)
All those have to be "myName".  It would be better to not have used upper case for names to begin with.

**Reply by TStone45** [reply]
My stoopid Camelcase ass..... on god you are a very kind man, greatly appreciate your time and efforts. 

I will update asap, working currently by day, coding by night.... not exactly the Dark Knight but, echoes....

**Reply by GaryAustin1** [reply]
@TStone45 It happens so much I have a meme I use...
![image](https://github.com/supabase/supabase/assets/54564956/973477b6-3caf-4d39-85c7-a767137a2eae)

**Comment by asiriPiyajanaka** [comment]
Here is how I did it using supabase triggers -  [https://stackoverflow.com/a/78582155/4350622](https://stackoverflow.com/a/78582155/4350622)
